compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/collectors/components/ErrorNodeDiagnosticCollectorComponent.kt

```diff
@@ -33,15 +33,28 @@ class ErrorNodeDiagnosticCollectorComponent(collector: AbstractDiagnosticCollect
     }
 
     override fun visitErrorNamedReference(errorNamedReference: FirErrorNamedReference, data: CheckerContext) {
-        val source = data.qualifiedAccesses.lastOrNull()?.takeIf {
+        val enclosingExpression = data.qualifiedAccesses.lastOrNull()?.takeIf {
             // Use the source of the enclosing FirQualifiedAccessExpression if it is exactly the call to the erroneous callee.
             it is FirQualifiedAccessExpression && it.calleeReference == errorNamedReference
-        }?.source ?: errorNamedReference.source ?: return
+        }
+        if (enclosingExpression != null &&
+            (enclosingExpression.dispatchReceiver.hasErrorType() || enclosingExpression.extensionReceiver.hasErrorType()
+                    || enclosingExpression.explicitReceiver.hasErrorType())
```

Comment:
Maybe I'm wrong, but I wonder previous two receivers cover the explicit receiver? Could you try removing the last condition?

Reply:
Consider the following. I think all three are needed.

```
fun String.foo() {
    "".length // explicit = FirConstExpression; extension = FirNoReceiver;  dispatch = null
    foo() // explicit = null; extension = FirThisReceiverExpression;  dispatch = FirNoReceiver
    length // explicit = null; extension = FirNoReceiver;  dispatch = FirThisReceiverExpression
}
```

Reply:
Ah NVM! I am confused by FirViewer output. Actually for the first case, the dispatchReceiver the same as explicitReceiver! Let me try removing the last one.

Reply:
OK, tried removing it and it looks like for unresolved cases the `explicitReceiver` is useful. For example `val x = random123.foo()` has an unresolved explicitReceiver and `FirNoReceiver` for dispatch and extenion.

Reply:
Oh.... I didn't expect unresolved cases! Good to know this.

Yeah, once everything is properly resolved, explicit receiver will point to either of extension receiver or dispatch receiver.